frame/collective/src/lib.rs

```
@@ -871,16 +910,18 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	/// Checks the length in storage via `storage::read` which adds an extra `size_of::<u32>() == 4`
 	/// to the length.
 	fn validate_and_get_proposal(
-		hash: &T::Hash,
+		proposal_bounded: &Bounded<<T as Config<I>>::Proposal>,
 		length_bound: u32,
 		weight_bound: Weight,
 	) -> Result<(<T as Config<I>>::Proposal, usize), DispatchError> {
-		let key = ProposalOf::<T, I>::hashed_key_for(hash);
-		// read the length of the proposal storage entry directly
-		let proposal_len =
-			storage::read(&key, &mut [0; 0], 0).ok_or(Error::<T, I>::ProposalMissing)?;
+		// read the length of the proposal by using peek
+		let (proposal, proposal_len) =
+			T::Preimages::peek(&proposal_bounded).map_err(|_| Error::<T, I>::ProposalMissing)?;
+		let proposal_len = proposal_len.unwrap_or_else(|| match proposal_bounded {
+			Bounded::Inline(data) => data.len() as u32,
+			_ => panic!("preimage should be inline, or len already given by peek; qed"),
```

You can just return a defensive error here.